File: infrastructures/ddd/python/src/infrastructure/repositories/impl/example_query_repository.py
"""
Example query repository implementation.
Repository implementations:
- Live in /infrastructure/repositories/impl/
- Implement repository interfaces
- Handle query operations
- Return ReadEntity directly (no BMO mapping in read path)
"""

from typing import List, Optional
from infrastructure.repositories.i_example_query_repository import IExampleQueryRepository
from infrastructure.persistence.read.example_read_entity import ExampleReadEntity


class ExampleQueryRepository(IExampleQueryRepository):
    """
    Query repository implementation for Example domain.
    
    This implementation:
    - Handles read operations (find, list, search)
    - Works with ReadEntity directly
    - Returns ReadEntity to service (no BMO in read path)
    - Optimized for query performance
    
    In a real application with MongoDB:
    ```python
    from motor.motor_asyncio import AsyncIOMotorClient, AsyncIOMotorDatabase
    
    class ExampleQueryRepository(IExampleQueryRepository):
        def __init__(self, p_database: AsyncIOMotorDatabase):
            self._collection = p_database["examples_view"]
        
        async def find_by_id_async(self, p_id: str) -> Optional[ExampleReadEntity]:
            doc = await self._collection.find_one({"_id": p_id})
            if doc is None:
                return None
            
            entity = ExampleReadEntity()
            entity.id = doc["_id"]
            entity.name = doc["name"]
            entity.description = doc["description"]
            entity.owner_id = doc["owner_id"]
            entity.owner_name = doc["owner_name"]
            entity.is_active = doc["is_active"]
            entity.display_name = doc["display_name"]
            entity.status_text = doc["status_text"]
            entity.created_at = doc["created_at"]
            entity.updated_at = doc["updated_at"]
            return entity
    ```
    """
    
    def __init__(self):
        """Initialize the query repository."""
        # In real implementation, inject database connection here
        self._in_memory_store: dict[str, ExampleReadEntity] = {}
    
    async def find_by_id_async(self, p_id: str) -> Optional[ExampleReadEntity]:
        """
        Finds an example by ID.
        
        Args:
            p_id: The ID to find
            
        Returns:
            The read entity or None if not found
        """
        # In real implementation, query database
        return self._in_memory_store.get(p_id)
    
    async def list_by_filter_async(
        self,
        p_skip: int = 0,
        p_take: int = 10
    ) -> List[ExampleReadEntity]:
        """
        Lists examples with pagination.
        
        Args:
            p_skip: Number of records to skip
            p_take: Number of records to take
            
        Returns:
            List of read entities
        """
        # In real implementation, query database with pagination
        all_entities = list(self._in_memory_store.values())
        return all_entities[p_skip:p_skip + p_take]

File: infrastructures/rest-ddd/java/src/main/java/com/example/restddd/infrastructure/repositories/impl/ExampleQueryRepository.java
/**
 * Example Query Repository Implementation
 * Handles read operations (SELECT).
 */

package com.example.restddd.infrastructure.repositories.impl;

import com.example.restddd.infrastructure.persistence.read.ExampleReadEntity;
import com.example.restddd.infrastructure.repositories.IQueryRepository;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;

public class ExampleQueryRepository implements IQueryRepository<ExampleReadEntity, String> {
    /**
     * Query Repository for Example entity.
     * 
     * Rules:
     * - Located in /infrastructure/repositories/impl/
     * - Suffix with 'Repository'
     * - Returns ReadEntity directly (NO BMO mapping)
     * - Optimized for query performance
     */

    // Add database connection/context here
    // private final DatabaseContext m_dbContext;

    public ExampleQueryRepository() {
    }

    @Override
    public CompletableFuture<ExampleReadEntity> findByIdAsync(String p_id) {
        return CompletableFuture.supplyAsync(() -> {
            // Query database for read entity
            // Document result = m_dbContext.getExamplesView()
            //     .find(Filters.eq("_id", p_id))
            //     .first();
            // 
            // if (result == null) {
            //     return null;
            // }
            // 
            // return new ExampleReadEntity(
            //     result.getString("_id"),
            //     result.getString("name"),
            //     result.getString("description"),
            //     result.getString("status"),
            //     result.getString("ownerId"),
            //     result.getString("ownerName"),
            //     result.getString("createdAt"),
            //     result.getString("updatedAt")
            // );

            // Placeholder
            return null;
        });
    }

    @Override
    public CompletableFuture<List<ExampleReadEntity>> listByFilterAsync(
        Map<String, Object> p_filter,
        int p_page,
        int p_pageSize
    ) {
        return CompletableFuture.supplyAsync(() -> {
            // Query database with filter and pagination
            // int skip = (p_page - 1) * p_pageSize;
            // List<Document> results = m_dbContext.getExamplesView()
            //     .find(buildFilter(p_filter))
            //     .skip(skip)
            //     .limit(p_pageSize)
            //     .into(new ArrayList<>());
            // 
            // return results.stream()
            //     .map(r -> new ExampleReadEntity(
            //         r.getString("_id"),
            //         r.getString("name"),
            //         r.getString("description"),
            //         r.getString("status"),
            //         r.getString("ownerId"),
            //         r.getString("ownerName"),
            //         r.getString("createdAt"),
            //         r.getString("updatedAt")
            //     ))
            //     .collect(Collectors.toList());

            // Placeholder
            return new ArrayList<>();
        });
    }

    @Override
    public CompletableFuture<Integer> countByFilterAsync(Map<String, Object> p_filter) {
        return CompletableFuture.supplyAsync(() -> {
            // Count documents matching filter
            // long count = m_dbContext.getExamplesView()
            //     .countDocuments(buildFilter(p_filter));
            // return (int) count;

            // Placeholder
            return 0;
        });
    }
}
